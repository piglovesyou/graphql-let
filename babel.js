module.exports = require('./dist/babel-plugin').default;
