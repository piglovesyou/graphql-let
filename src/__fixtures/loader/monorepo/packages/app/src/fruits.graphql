query GetFruits {
    fruits {
        id
        name
        flavor
    }
}
