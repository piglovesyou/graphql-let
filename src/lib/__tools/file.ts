import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
