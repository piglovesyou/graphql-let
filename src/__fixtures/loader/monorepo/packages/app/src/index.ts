import { useGetFruitsQuery } from './fruits.graphql';

export function App() {
  const { data, loading } = useGetFruitsQuery();

  if (loading) {
    return 'Loading...';
  }

  return JSON.stringify(data, null, 2);
}
