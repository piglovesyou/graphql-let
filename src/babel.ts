import { types, ConfigAPI, PluginObj, NodePath } from '@babel/core';
import * as t from '@babel/types';
import { relative, dirname, join as pathJoin } from 'path';
import { declare } from '@babel/helper-plugin-utils';
import doSync from 'do-sync';
// import createDebug from 'debug';
import { readFileSync } from './lib/file';
import { GqlCodegenContext, GqlCompileArgs } from './lib/gql-compile';
import { createHash } from './lib/hash';
import { ConfigTypes, loadConfigSync } from './lib/config';
import { shouldGenResolverTypes } from './lib/resolver-types';

// // TODO: Utilize it
// const debug = createDebug('graphql-let/babel');

const {
  // cloneDeep,
  isIdentifier,
  // isMemberExpression,
  isImportDefaultSpecifier,
  // variableDeclaration,
  // variableDeclarator,
  // memberExpression,
  // callExpression,
  identifier,
  importDeclaration,
  importNamespaceSpecifier,
  valueToNode,
} = types;

const gqlCompileSync = doSync(
  async ({
    hostDirname,
    ...gqlCompileArgs
  }: GqlCompileArgs & { hostDirname: string }) => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { join } = require('path');
    const modulePath = join(hostDirname, '../dist/lib/gql-compile');
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { gqlCompile } = require(modulePath);
    return await gqlCompile(gqlCompileArgs);
  },
);

export type BabelOptions = {
  configFilePath?: string;
  importName?: string;
  onlyMatchImportSuffix?: boolean;
};

const ensureConfig = (() => {
  let config: ConfigTypes | null = null;
  let schemaHash: string | null = null;
  return (cwd: string, configFilePath: string): [ConfigTypes, string] => {
    if (config && schemaHash) {
      return [config, schemaHash];
    }
    const [_config, configHash] = loadConfigSync(cwd, configFilePath);
    // TODO: refactor with create-codegen-opts.ts
    config = {
      ..._config,
      config: {
        withHOC: false, // True by default
        withHooks: true, // False by default
        ..._config.config,
      },
    };

    schemaHash = configHash;
    if (shouldGenResolverTypes(config)) {
      const fileSchema = config.schema as string;
      const schemaFullPath = pathJoin(cwd, fileSchema);
      const content = readFileSync(schemaFullPath);
      schemaHash = createHash(schemaHash + content);
    }
    return [config, schemaHash];
  };
})();

// With all my respect, I cloned the source from
// https://github.com/gajus/babel-plugin-graphql-tag/blob/master/src/index.js
const configFunction = (
  api: ConfigAPI,
  options: BabelOptions,
): PluginObj<any> => {
  api.assertVersion(7);
  const {
    configFilePath = '',
    importName = 'graphql-let',
    onlyMatchImportSuffix = false,
    // strip = false,
  } = options;

  return {
    visitor: {
      Program(programPath: NodePath<t.Program>, state: any) {
        const { cwd } = state;
        const sourceFullPath = state.file.opts.filename;
        const sourceRelPath = relative(cwd, sourceFullPath);

        const [graphqlLetConfig, schemaHash] = ensureConfig(
          cwd,
          configFilePath,
        );
        const cacheRelDir = graphqlLetConfig.cacheDir;

        const tagNames: string[] = [];
        const pendingDeletion: {
          defaultSpecifier:
            | t.ImportSpecifier
            | t.ImportDefaultSpecifier
            | t.ImportNamespaceSpecifier;
          path: NodePath<t.ImportDeclaration>;
        }[] = [];
        const gqlCallExpressionPaths: [
          NodePath<t.CallExpression>,
          string,
        ][] = [];
        let hasError = false;

        programPath.traverse({
          ImportDeclaration(path: NodePath<t.ImportDeclaration>) {
            const pathValue = path.node.source.value;
            if (
              onlyMatchImportSuffix
                ? pathValue.endsWith(importName)
                : pathValue === importName
            ) {
              const defaultSpecifier = path.node.specifiers.find(
                (specifier) => {
                  return isImportDefaultSpecifier(specifier);
                },
              );

              if (defaultSpecifier) {
                tagNames.push(defaultSpecifier.local.name);
                pendingDeletion.push({
                  defaultSpecifier,
                  path,
                });
              }
            }
          },
          // TODO: Allow TemplateLiteralCall with warning:
          // "transpile is ok, but it can't finish type injection"
          CallExpression(path: NodePath<t.CallExpression>) {
            if (
              tagNames.some((name) => {
                return isIdentifier(path.node.callee, { name });
              })
            ) {
              try {
                const args = path.get('arguments');
                if (args.length !== 1)
                  throw new Error(
                    `The argument must be a single string value.`,
                  );
                let value = '';
                path.traverse({
                  TemplateLiteral(path: NodePath<t.TemplateLiteral>) {
                    if (path.node.quasis.length !== 1)
                      throw new Error(
                        `TemplateLiteral of the argument must not contain arguments.`,
                      );
                    value = path.node.quasis[0].value.raw;
                  },
                  StringLiteral(path: NodePath<t.StringLiteral>) {
                    value = path.node.value;
                  },
                });
                if (!value) throw new Error('never');
                gqlCallExpressionPaths.push([path, value]);
              } catch (error) {
                // eslint-disable-next-line no-console
                console.error('error', error);
                hasError = true;
              }
            }
          },
        });

        // TODO: Handle error

        if (!gqlCallExpressionPaths.length) return;

        const rv: GqlCodegenContext = gqlCompileSync({
          hostDirname: __dirname,
          cwd,
          sourceRelPath,
          gqlContents: gqlCallExpressionPaths.map(([, value]) => value),
          cacheRelDir, // TODO: Include in config
          schemaHash,
          config: graphqlLetConfig,
        });
        if (gqlCallExpressionPaths.length !== rv.length)
          throw new Error('what');

        for (const [
          i,
          [callExpressionPath],
        ] of gqlCallExpressionPaths.entries()) {
          const { gqlContentHash, tsxFullPath } = rv[i]!;
          const tsxRelPathFromSource =
            './' + relative(dirname(sourceFullPath), tsxFullPath);

          const localVarName = `V${gqlContentHash}`;

          const importNode = importDeclaration(
            [importNamespaceSpecifier(identifier(localVarName))],
            valueToNode(tsxRelPathFromSource),
          );

          programPath.unshiftContainer('body', importNode);
          callExpressionPath.replaceWithSourceString(localVarName);
        }

        // Only delete import statement or specifier when there is no error
        if (!hasError) {
          for (const {
            defaultSpecifier,
            path: pathForDeletion,
          } of pendingDeletion) {
            if (pathForDeletion.node.specifiers.length === 1) {
              pathForDeletion.remove();
            } else {
              // TODO what's going on
              pathForDeletion.node.specifiers = pathForDeletion.node.specifiers.filter(
                (specifier) => {
                  return specifier !== defaultSpecifier;
                },
              );
            }
          }
        }
      },
    },
  };
};

export default declare(configFunction);
