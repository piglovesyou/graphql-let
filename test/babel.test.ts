import { deepStrictEqual } from 'assert';
import globby from 'globby';
import { join } from 'path';
import runner from '@babel/helper-transform-fixture-test-runner';
import { readFile, rimraf } from './__tools/file';
import { clearExecContext } from '../src/babel';

const cwd = join(__dirname, '__fixtures/babel');
const rel = (path: string) => join(cwd, path);

beforeAll(async () => {
  await rimraf(join(cwd, '**/node_modules'));
});

// We cache "config", etc. for production.
// Here we traverse multiple fixture projects, so erase them for each.
beforeEach(() => clearExecContext());

runner(cwd, 'gql', {}, { sourceType: 'unambiguous' });

test(`Generated files are okay`, async () => {
  const expectedFiles = [
    'fixtures/basic/node_modules/@types/graphql-let/index.d.ts',
    'fixtures/basic/node_modules/@types/graphql-let/input-3daf5b4311465bd8566d77d548e135871cc11646.d.ts',
    'fixtures/basic/node_modules/@types/graphql-let/store.json',
    'fixtures/basic/node_modules/graphql-let/__generated__/input-3daf5b4311465bd8566d77d548e135871cc11646.tsx',
    'fixtures/tagged-template-call/node_modules/@types/graphql-let/index.d.ts',
    'fixtures/tagged-template-call/node_modules/@types/graphql-let/input-3daf5b4311465bd8566d77d548e135871cc11646.d.ts',
    'fixtures/tagged-template-call/node_modules/@types/graphql-let/store.json',
    'fixtures/tagged-template-call/node_modules/graphql-let/__generated__/input-3daf5b4311465bd8566d77d548e135871cc11646.tsx',
  ];
  const files = await globby('fixtures/*/node_modules/**', { cwd });

  deepStrictEqual(files.sort(), expectedFiles);

  const contents = await Promise.all(
    files.map((file) => {
      return readFile(rel(file)).then((content) => [file, content]);
    }),
  );

  for (const [file, content] of contents) expect(content).toMatchSnapshot(file);
});
