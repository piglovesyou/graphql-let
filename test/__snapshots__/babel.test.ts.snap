// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated files are okay: fixtures/*/node_modules/** 1`] = `
Array [
  "fixtures/basic/node_modules/@types/graphql-let/index.d.ts",
  "fixtures/basic/node_modules/@types/graphql-let/input-1ffee34aad94b8da4e33faf385ded9238d7affef.d.ts",
  "fixtures/basic/node_modules/@types/graphql-let/store.json",
  "fixtures/basic/node_modules/graphql-let/__generated__/input-1ffee34aad94b8da4e33faf385ded9238d7affef.tsx",
  "fixtures/tagged-template-call/node_modules/@types/graphql-let/index.d.ts",
  "fixtures/tagged-template-call/node_modules/@types/graphql-let/input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.d.ts",
  "fixtures/tagged-template-call/node_modules/@types/graphql-let/store.json",
  "fixtures/tagged-template-call/node_modules/graphql-let/__generated__/input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.tsx",
  "fixtures/with-fragments/node_modules/@types/graphql-let/index.d.ts",
  "fixtures/with-fragments/node_modules/@types/graphql-let/input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.d.ts",
  "fixtures/with-fragments/node_modules/@types/graphql-let/store.json",
  "fixtures/with-fragments/node_modules/graphql-let/__generated__/input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.tsx",
]
`;

exports[`Generated files are okay: fixtures/basic/node_modules/@types/graphql-let/index.d.ts 1`] = `
"import T1ffee34aad94b8da4e33faf385ded9238d7affef from './input-1ffee34aad94b8da4e33faf385ded9238d7affef.d.ts';
export default function gql(gql: \`query Viewer { viewer { name } }\`): T1ffee34aad94b8da4e33faf385ded9238d7affef.__AllExports;
"
`;

exports[`Generated files are okay: fixtures/basic/node_modules/@types/graphql-let/input-1ffee34aad94b8da4e33faf385ded9238d7affef.d.ts 1`] = `
"/* 1ffee34aad94b8da4e33faf385ded9238d7affef
 * This file is automatically generated by graphql-let. */

import * as Apollo from '@apollo/client';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
  [key: string]: unknown;
}> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */

export declare type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};
export declare type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};
export declare type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};
export declare type ViewerQueryVariables = Exact<{
  [key: string]: never;
}>;
export declare type ViewerQuery = ({
  __typename?: 'Query';
} & {
  viewer?: Maybe<({
    __typename?: 'User';
  } & Pick<User, 'name'>)>;
});
export declare const ViewerDocument: Apollo.DocumentNode;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export declare function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryResult<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryTuple<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export declare type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export declare type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;
export declare type __AllExports = {
  Maybe: Maybe;
  Exact: Exact;
  Scalars: Scalars;
  Query: Query;
  User: User;
  ViewerQueryVariables: ViewerQueryVariables;
  ViewerQuery: ViewerQuery;
  ViewerDocument: ViewerDocument;
  useViewerQuery: typeof useViewerQuery;
  useViewerLazyQuery: typeof useViewerLazyQuery;
  ViewerQueryHookResult: ViewerQueryHookResult;
  ViewerLazyQueryHookResult: ViewerLazyQueryHookResult;
  ViewerQueryResult: ViewerQueryResult;
};"
`;

exports[`Generated files are okay: fixtures/basic/node_modules/@types/graphql-let/store.json 1`] = `
"{
  \\"input.tsx\\": {
    \\"1ffee34aad94b8da4e33faf385ded9238d7affef\\": [
      \\"input-1ffee34aad94b8da4e33faf385ded9238d7affef.d.ts\\",
      \\"query Viewer { viewer { name } }\\"
    ]
  }
}"
`;

exports[`Generated files are okay: fixtures/basic/node_modules/graphql-let/__generated__/input-1ffee34aad94b8da4e33faf385ded9238d7affef.tsx 1`] = `
"/* 1ffee34aad94b8da4e33faf385ded9238d7affef
 * This file is automatically generated by graphql-let. */

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};

export type ViewerQueryVariables = Exact<{ [key: string]: never; }>;


export type ViewerQuery = (
  { __typename?: 'Query' }
  & { viewer?: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'name'>
  )> }
);


export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    name
  }
}
    \`;

/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */
export function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
        return Apollo.useQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
      }
export function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
          return Apollo.useLazyQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
        }
export type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;"
`;

exports[`Generated files are okay: fixtures/tagged-template-call/node_modules/@types/graphql-let/index.d.ts 1`] = `
"import T98fe64b5c1bccab885117fbeac3ea597d4f7e02d from './input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.d.ts';
export default function gql(gql: \`
    query Viewer {
        viewer {
            name
        }
    }
\`): T98fe64b5c1bccab885117fbeac3ea597d4f7e02d.__AllExports;
"
`;

exports[`Generated files are okay: fixtures/tagged-template-call/node_modules/@types/graphql-let/input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.d.ts 1`] = `
"/* 98fe64b5c1bccab885117fbeac3ea597d4f7e02d
 * This file is automatically generated by graphql-let. */

import * as Apollo from '@apollo/client';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
  [key: string]: unknown;
}> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */

export declare type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};
export declare type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};
export declare type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};
export declare type ViewerQueryVariables = Exact<{
  [key: string]: never;
}>;
export declare type ViewerQuery = ({
  __typename?: 'Query';
} & {
  viewer?: Maybe<({
    __typename?: 'User';
  } & Pick<User, 'name'>)>;
});
export declare const ViewerDocument: Apollo.DocumentNode;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export declare function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryResult<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryTuple<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export declare type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export declare type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;
export declare type __AllExports = {
  Maybe: Maybe;
  Exact: Exact;
  Scalars: Scalars;
  Query: Query;
  User: User;
  ViewerQueryVariables: ViewerQueryVariables;
  ViewerQuery: ViewerQuery;
  ViewerDocument: ViewerDocument;
  useViewerQuery: typeof useViewerQuery;
  useViewerLazyQuery: typeof useViewerLazyQuery;
  ViewerQueryHookResult: ViewerQueryHookResult;
  ViewerLazyQueryHookResult: ViewerLazyQueryHookResult;
  ViewerQueryResult: ViewerQueryResult;
};"
`;

exports[`Generated files are okay: fixtures/tagged-template-call/node_modules/@types/graphql-let/store.json 1`] = `
"{
  \\"input.tsx\\": {
    \\"98fe64b5c1bccab885117fbeac3ea597d4f7e02d\\": [
      \\"input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.d.ts\\",
      \\"\\\\n    query Viewer {\\\\n        viewer {\\\\n            name\\\\n        }\\\\n    }\\\\n\\"
    ]
  }
}"
`;

exports[`Generated files are okay: fixtures/tagged-template-call/node_modules/graphql-let/__generated__/input-98fe64b5c1bccab885117fbeac3ea597d4f7e02d.tsx 1`] = `
"/* 98fe64b5c1bccab885117fbeac3ea597d4f7e02d
 * This file is automatically generated by graphql-let. */

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};

export type ViewerQueryVariables = Exact<{ [key: string]: never; }>;


export type ViewerQuery = (
  { __typename?: 'Query' }
  & { viewer?: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'name'>
  )> }
);


export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    name
  }
}
    \`;

/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */
export function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
        return Apollo.useQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
      }
export function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
          return Apollo.useLazyQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
        }
export type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;"
`;

exports[`Generated files are okay: fixtures/with-fragments/node_modules/@types/graphql-let/index.d.ts 1`] = `
"import T3e81ed6254d58faa4d4bb8a12568fa59509c0acb from './input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.d.ts';
export default function gql(gql: \`
  # import Partial from './partial.graphql'
  query Viewer {
    viewer {
      ...Partial
    }
  }
\`): T3e81ed6254d58faa4d4bb8a12568fa59509c0acb.__AllExports;
"
`;

exports[`Generated files are okay: fixtures/with-fragments/node_modules/@types/graphql-let/input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.d.ts 1`] = `
"/* 3e81ed6254d58faa4d4bb8a12568fa59509c0acb
 * This file is automatically generated by graphql-let. */

import * as Apollo from '@apollo/client';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
  [key: string]: unknown;
}> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */

export declare type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};
export declare type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};
export declare type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};
export declare type ViewerQueryVariables = Exact<{
  [key: string]: never;
}>;
export declare type ViewerQuery = ({
  __typename?: 'Query';
} & {
  viewer?: Maybe<({
    __typename?: 'User';
  } & PartialFragment)>;
});
export declare type PartialFragment = ({
  __typename?: 'User';
} & Pick<User, 'id' | 'name'>);
export declare const PartialFragmentDoc: Apollo.DocumentNode;
export declare const ViewerDocument: Apollo.DocumentNode;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export declare function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryResult<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>): Apollo.QueryTuple<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export declare type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export declare type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;
export declare type __AllExports = {
  Maybe: Maybe;
  Exact: Exact;
  Scalars: Scalars;
  Query: Query;
  User: User;
  ViewerQueryVariables: ViewerQueryVariables;
  ViewerQuery: ViewerQuery;
  PartialFragment: PartialFragment;
  PartialFragmentDoc: PartialFragmentDoc;
  ViewerDocument: ViewerDocument;
  useViewerQuery: typeof useViewerQuery;
  useViewerLazyQuery: typeof useViewerLazyQuery;
  ViewerQueryHookResult: ViewerQueryHookResult;
  ViewerLazyQueryHookResult: ViewerLazyQueryHookResult;
  ViewerQueryResult: ViewerQueryResult;
};"
`;

exports[`Generated files are okay: fixtures/with-fragments/node_modules/@types/graphql-let/store.json 1`] = `
"{
  \\"input.tsx\\": {
    \\"3e81ed6254d58faa4d4bb8a12568fa59509c0acb\\": [
      \\"input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.d.ts\\",
      \\"\\\\n  # import Partial from './partial.graphql'\\\\n  query Viewer {\\\\n    viewer {\\\\n      ...Partial\\\\n    }\\\\n  }\\\\n\\"
    ]
  }
}"
`;

exports[`Generated files are okay: fixtures/with-fragments/node_modules/graphql-let/__generated__/input-3e81ed6254d58faa4d4bb8a12568fa59509c0acb.tsx 1`] = `
"/* 3e81ed6254d58faa4d4bb8a12568fa59509c0acb
 * This file is automatically generated by graphql-let. */

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};

export type ViewerQueryVariables = Exact<{ [key: string]: never; }>;


export type ViewerQuery = (
  { __typename?: 'Query' }
  & { viewer?: Maybe<(
    { __typename?: 'User' }
    & PartialFragment
  )> }
);

export type PartialFragment = (
  { __typename?: 'User' }
  & Pick<User, 'id' | 'name'>
);

export const PartialFragmentDoc = gql\`
    fragment Partial on User {
  id
  name
}
    \`;
export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    ...Partial
  }
}
    \${PartialFragmentDoc}\`;

/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */
export function useViewerQuery(baseOptions?: Apollo.QueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
        return Apollo.useQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
      }
export function useViewerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
          return Apollo.useLazyQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
        }
export type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export type ViewerQueryResult = Apollo.QueryResult<ViewerQuery, ViewerQueryVariables>;"
`;
